cinder/volume/drivers/vmware/vmdk.py

```diff
@@ -451,23 +452,74 @@ def _get_volume_stats(self):
                     self.session.vim.service_content.rootFolder,
                     [vim_util.build_recursive_traversal_spec(client_factory)]))
 
-        global_capacity = 0
-        global_free = 0
-        # If there are no datastores, then object specs are empty
-        # we can't query vcenter with empty object specs, or we'll
-        # get errors.
-        if object_specs:
-            prop_spec = vim_util.build_property_spec(
-                client_factory, 'Datastore', ['summary'])
-            filter_spec = vim_util.build_property_filter_spec(
-                client_factory, prop_spec, object_specs)
-            options = client_factory.create('ns0:RetrieveOptions')
-            max_objects = self.configuration.vmware_max_objects_retrieval
-            options.maxObjects = max_objects
-            result = self.session.vim.RetrievePropertiesEx(
-                self.session.vim.service_content.propertyCollector,
-                specSet=[filter_spec],
-                options=options)
+        prop_spec = vim_util.build_property_spec(
+            client_factory, 'Datastore', ['summary'])
+        filter_spec = vim_util.build_property_filter_spec(
+            client_factory, prop_spec, object_specs)
+        options = client_factory.create('ns0:RetrieveOptions')
+        options.maxObjects = self.configuration.vmware_max_objects_retrieval
+        result = self.session.vim.RetrievePropertiesEx(
+            self.session.vim.service_content.propertyCollector,
+            specSet=[filter_spec],
+            options=options)
+
+        return (result, datastore_profiles)
+
+    @utils.trace
+    def _get_volume_stats(self):
+        backend_name = self.configuration.safe_get('volume_backend_name')
+        if not backend_name:
+            backend_name = self.__class__.__name__
+
+        location_info = '%(driver_name)s:%(vcenter)s' % {
+            'driver_name': LOCATION_DRIVER_NAME,
+            'vcenter': self.session.vim.service_content.about.instanceUuid}
+        pools = []
+        reserved_percentage = self.configuration.reserved_percentage
+        max_over_subscription_ratio = self.configuration.safe_get(
+            'max_over_subscription_ratio')
+
+        data = {'volume_backend_name': backend_name,
+                'vendor_name': 'VMware',
+                'driver_version': self.VERSION,
+                'storage_protocol': 'vmdk',
+                'location_info': location_info,
+                'connection_capabilities': self._get_connection_capabilities(),
+                }
+
+        result, datastore_profiles = self._collect_backend_stats()
+        if self.configuration.vmware_datastores_as_pools:
+            while True:
+                for ds in result.objects:
```

Comment:
I know this pattern of explicitly calling `ContinueRetrievePropertiesEx` predates your changes, but for me it looks it calls for the use of `vim_utill.WithRetrieval` [Source](https://github.com/openstack/oslo.vmware/blob/stable/train/oslo_vmware/vim_util.py#L406-L408).
Or at least [continue_retrieval](https://github.com/openstack/oslo.vmware/blob/8b49bd55e5f2274d14745e95241b35f7b1c9e853/oslo_vmware/vim_util.py#L384)

I am also a bit unconvinced by having the `_collect_backend_stats` return only a partial result, on which the caller has to call continue results over, instead of yielding an iterator to the results themselves.

Reply:
_collect_backend_stats was created to ensure that in both pools enabled and disabled, the same code ran to start the query to get the stats.  There is no need to change the logic of it.

Reply:
The point is that the code is leaking resources on the vcenter, because it is not calling CancelRetrieval in case of an exception. `WithRetrieval` has encapsulated that plus `continue_retrieval`

Reply:
This is outside the scope of the purpose of this patch.

Reply:
Point taken

Reply:
Let's separate the two things. What is the right way, and what are the possible ways.

You are stating it is not possible, but as I read it, the following four lines instead of the commented ones should suffice:
```
        if (self.configuration.vmware_datastores_as_pools and
                self._is_pool_outdated_for_volume(volume)):
                    raise exception.ConnectorRejected(
                          reason="Non-c %s" % ', '.join(missing))
```
(Yes, that probably should be rather in `_initialise_connection`)

For standard volume attachments, we go through `_initialise_connection`, for backups (backup & restore), we also create a connection for brick, snapshot of empty volumes don't matter as they are a no-op and do not call `_create_backing`. And is known to work, as we do it that way in #50.

So, what am I missing? What doesn't work there?

Assuming it is possible, and we discuss design choices:
> The user didn't want a migrate, they wanted an attach or a clone or a backup.

We are looking here at different levels of abstractions. It isn't about what the user asks over the rest-api, but how the cinder volume manager communicates with the driver.

From my point of view, one purpose of abstractions is that they give us the liberty of choosing how to achieve the desired result, without the caller having to care. It was the path chosen in #50. But here the abstraction is being leaky as it could not be realised in constant time.

Now, exactly because there is no backing, the volume manager can ask the driver to move the volume from one cinder driver instance to another (both being the same instance, most likely), and realise the migration (of the record/responsibility) in constant time.

Yes, there are downsides of doing that, but I would like to list the downsides of the approaches and make a decision based on that.

Another thing to consider: As I understand it, that code is purely for migration purposes. Another option might be to move it out of the driver completely and realise the migration with a script.